interfaces/cython/cantera/test/test_utils.py

```diff
@@ -0,0 +1,117 @@
+import numpy as np
+
+import cantera as ct
+from . import utilities
+from .utilities import unittest
+
+from cantera._cantera import _py_to_any_to_py
+
+class TestPyToAnyValue(utilities.CanteraTest):
+
+    def check_conversion(self, value, check_type=None):
+        out, held_type = _py_to_any_to_py(value)
+        self.assertEqual(out, value)
+        if check_type is not None:
+            self.assertEqual(held_type, check_type)
+
+    def check_inexact_conversion(self, value, check_type=None):
+        out, held_type = _py_to_any_to_py(value)
+        if isinstance(value, np.ndarray):
+            self.assertEqual(out, value.tolist())
+        else:
+            self.assertEqual(out, list(value))
+        if check_type is not None:
+            self.assertEqual(held_type, check_type)
+
+    def check_raises(self, value, ee, regex):
+        with self.assertRaisesRegex(ee, regex):
+            _py_to_any_to_py(value)
+
+    def test_none(self):
+        out, held_type = _py_to_any_to_py(None)
+        self.assertEqual(out, None)
+        self.assertEqual(held_type, "void")
+
+    def test_set(self):
+        # Sets are converted to lists
+        self.check_inexact_conversion({'a', 'b'}, 'vector<string>')
+
+    def test_empty_list(self):
+        self.check_conversion([])
+
+    def test_empty_ndarray(self):
+        self.check_inexact_conversion(np.ndarray((0,)))
+
+    def test_empty_dict(self):
+        self.check_conversion({})
+
+    def test_scalar_string(self):
+        self.check_conversion('spam', 'string')
+
+    def test_scalar_int(self):
+        self.check_conversion(3, 'long int')
+
+    def test_scalar_float(self):
+        self.check_conversion(3.1415, 'double')
+
+    def test_scalar_bool(self):
+        self.check_conversion(True, 'bool')
+
+    def test_list_string(self):
+        self.check_conversion(['spam', 'eggs'], 'vector<string>')
+
+    def test_list_int(self):
+        self.check_conversion([1, 2, 3], 'vector<long int>')
+
+    def test_list_float(self):
+        self.check_conversion([1., 2., 3.], 'vector<double>')
+
+    def test_list_bool(self):
+        self.check_conversion([True, False], 'vector<bool>')
+
+    def test_list_various(self):
+        self.check_conversion([True, 'spam', 3, 4., {'foo': 'bar'}],
+                              'vector<AnyValue>')
+
+    def test_tuple(self):
+        self.check_inexact_conversion((True, 'spam', 3, 4.), 'vector<AnyValue>')
+
+    def test_ndarray1(self):
+        self.check_inexact_conversion(np.random.randn(10), 'vector<double>')
+
+    def test_ndarray2(self):
+        self.check_inexact_conversion(np.random.randn(3, 2), 'vector<vector<double>>')
+
+    def test_ndarray3(self):
+        # Each inner AnyValue holds a vector<vector<double>>
+        self.check_inexact_conversion(np.random.randn(3, 2, 4), 'vector<AnyValue>')
```

Comment:
I had blocked these conversions as 'not implemented', as the logical extension of `test_ndarray2` would be to produce a `vector<vector<vector<double>>>` here. It won't be felt in the Python API, but any (future) C++ implementation would be affected.

Reply:
What I was aiming for was consistency with the YAML to `AnyMap` conversion: for an n-dimensional array, the last two dimensions are converted to `vector<vector<double>>`, and then stored in nested `vector<AnyValue>`.

In C++, we have to draw the line somewhere, since you can't generate these new types dynamically. Based on the additional complexity it leads to and the limited ways in which it's used, I'm sort of wondering if we should just do this for the last dimension only, both here and when reading from YAML.

Reply:
I completely agree that the line has to be drawn (hence my `NotImplementedError` in lieu of an implementation). To my knowledge, it isn't used at the moment, and anyone needing it in the future would have to be sufficiently familiar with C++ to make a decision on how to implement this based on an actual use case.

Reply:
Decided to 'hide' my last comment as it really doesn't address this properly. Since there's a YAML precedent, I have to reassess - I hadn't noticed that possibility in the reverse `anyvalue_to_python`, which I why I decided to leave this unimplemented.

Regarding the specific implementation: `vector<vector<double>>` isn't ideal, but it maps to 2D arrays that are commonly used. Another alternative would be to use a boost array (I guess), although I'm not sure I want to go there. So I'm happy to stick with `<vector<vector<...>>` as I cannot think of any use cases at the moment.